.github/scripts/transform_dbt_test_results.py

```diff
@@ -0,0 +1,1082 @@
+                possible_fieldnames
+            )
+        )
+
+
+def main() -> None:
+    """Entrypoint to this script. Parses dbt test results and writes artifacts
+    to the output directory with metadata about tests."""
+    try:
+        run_results_filepath = sys.argv[1]
+    except IndexError:
+        run_results_filepath = os.path.join("target", "run_results.json")
+
+    try:
+        manifest_filepath = sys.argv[2]
+    except IndexError:
+        manifest_filepath = os.path.join("target", "manifest.json")
+
+    date_today = datetime.datetime.today().strftime("%Y-%m-%d")
+    try:
+        output_directory = sys.argv[3]
+    except IndexError:
+        output_directory = f"qc_test_results_{date_today}"
+
+    test_cache_path = get_test_cache_path(
+        run_results_filepath,
+        manifest_filepath,
+    )
+
+    if os.path.isfile(test_cache_path):
+        print(f"Loading test results from cache at {test_cache_path}")
+        test_categories = get_test_categories_from_file(test_cache_path)
+    else:
+        print(
+            f"Test cache not found at {test_cache_path}, loading test results "
+            "from Athena"
+        )
+        test_categories = get_test_categories_from_athena(
+            run_results_filepath,
+            manifest_filepath,
+        )
+        print(f"Saving test results to the cache at {test_cache_path}")
+        save_test_categories_to_file(test_categories, test_cache_path)
+
+    print("Generating the output workbook")
+    # It's important to use a write-only workbook here because otherwise
+    # the metadata required to store cell info about a large number of failing
+    # tests can cause the process to run out of memory
+    workbook = openpyxl.Workbook(write_only=True)
+    for test_category in test_categories:
+        print(f"Adding sheet for {test_category.category}")
+        test_category.add_to_workbook(workbook)
+
+    pathlib.Path(output_directory).mkdir(exist_ok=True)
+    workbook_filepath = os.path.join(
+        output_directory, f"qc_test_failures_{date_today}.xlsx"
+    )
+    workbook.save(workbook_filepath)
+    print(f"Output workbook saved to {workbook_filepath}")
+
+    # Generate and save metadata tables as parquet
+    test_run_metadata = TestRunMetadata.create(run_results_filepath)
+    test_run_result_metadata_list = TestRunResultMetadata.create_list(
+        test_categories, run_results_filepath
+    )
+    for metadata_list, tablename, partition_cols in [
+        ([test_run_metadata], "test_run", ["run_id"]),
+        (test_run_result_metadata_list, "test_run_result", ["run_id"]),
+    ]:
+        table = pyarrow.Table.from_pylist(
+            [dataclasses.asdict(meta_obj) for meta_obj in metadata_list]
+        )
+        metadata_root_path = os.path.join(
+            output_directory, "metadata", tablename
+        )
+        pyarrow.parquet.write_to_dataset(
+            table, metadata_root_path, partition_cols
+        )
```

Comment:
I'm not 100% sure about partitioning by run ID. It seems like an obvious choice, but I suspect it's going to lead to small file sizes for each `test_run` and `test_run_result` object, which we know creates overhead in Athena queries. It would probably be more performant from a query perspective if each script run loaded and then saved a couple shared parquet files in S3, each one partitioned by year, but from an engineering perspective I have a strong preference for small immutable files over big mutable files. My thinking is that it's easier to move forward with small immutable files and then refactor to big mutable files if we find that latency due to query overhead is severe, but I'm open to other opinions.

Reply:
Whoops, didn't see this until scrolling down, but see my earlier comment: we're on the same page. We really don't need ultra-performant queries, and lots of small immutable files should be way easier to deal with than some big mutable ones.